### PR TITLE
refactor how we rearrange fds for stdin/stdout/stderr

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/basic/fd-util.c [hunk 1]
```diff
@@ -615,3 +615,169 @@ int fd_move_above_stdio(int fd) {
...
+         * failure! Thus, callers should assume that when this function returns the input fds are invalidated.
+         *
+         * Note that when this function fails stdin/stdout/stderr might remain half set up!
+         *
+         * O_CLOEXEC is turned off for all three file descriptors (which is how it should be for
+         * stdin/stdout/stderr). */
+
+        if (original_input_fd == STDIN_FILENO)
+                input_fd = -1; /* Nothing to do, fd is already at the right place */
+        else {
```
Comment (keszybz) [inline src/basic/fd-util.c]: Hmm, there's essentially the same code repeated three times. I think it'd be much better to add `rearrange_one_fd` for the part in line 636–662.

File: src/basic/fd-util.c [hunk 2]
```diff
@@ -615,3 +615,169 @@ int fd_move_above_stdio(int fd) {
...
+
+                        error_fd = new_error_fd = copy;
+                }
+        }
+
+        /* At this point we now have the fds to us in input_fd, output_fd and error_fd, and they are all above the
+         * stdio range, so that we have freedom to move them around. If the fds already were at the right places then
+         * the specific fds are -1. Let's now move them to the right places. This is the point of no return. */
+
+        if (input_fd < 0) {
```
Comment (keszybz) [inline src/basic/fd-util.c]: Here too, why not make this a loop over an array of fds? Repeating the same code three times is error prone.

Comment (poettering) [reply src/basic/fd-util.c]: this one is different, indeed.

**Comment by poettering** [comment]
Force pushed a new version. No changes, except for the actual function which I reworked entirely. PTAL

**Pull request merged**